Supreme Props Mod FAF integration

badbwoi opened the pull request:
Here is the workflow on how I integrated an improved version of the "Supreme Props Mod" into FAF.

Infos about the original mod can be found here: http://forums.faforever.com/viewtopic.php?f=41&t=15934#p161087
BUG & LOD FIXES are integrated like in the original mod (video presentation: https://tinyurl.com/y6zteb8r)
The reclaim values remain mostly untouched, except I considered them as BUG FIXES. The props I touched reclaim wise are highlighted in the picture, see here: http://forums.faforever.com/viewtopic.php?f=41&t=15934&p=171871#p171312 The applied values match "mass_value A", except "D", special "M" and "map elements".

CHANGES:
1) FILE INPUT & FILE FIXES (manual)
- added 234 original FA env "_prop.bp" files to FAF
- replaced '# ' with '-- ' comments
- removed white spaces, tabs to spaces, corrected formatting
- fixed structure in \env\Geothermal\Props\Rocks\GeoRock01_prop.bp
- added new 'Oak01_Group2_LOD2_albedo.dds' to fix shade error!

2) READ ALL CURRENT BP DATA & WRITE CSV (python)
- fix values like SizeZ = .0 --> 0.0
- excluded Seraphim II, marker, emitter and other unsuited props

3) FIXES AND CHANGES IN CSV (manual)
- apply Supreme Props LOD where it was needed
- fix wrong or none LOD 3 (including shadername3 and adapted shadername2 if needed)
- fix SIZE where it is off (too huge block e.g. Crystalline; to small block e.g. Aeon, Cybran & Seraphim ). This also improves the collision boxes (check in game on fa_props map with ctrl+shift+alt+c). Some conflict could not be resolved due to the prop origin not being in the (volume) center
- fix RECLAIM VALUES where it was needed e.g. Structures (had no value), most of broken 300 mass props like small crystals (this allows map maker to use the Crystalline props more often), Aeon, Cybran, Seraphim & UEF and some props that were clearly off. See info and change pic: http://forums.faforever.com/viewtopic.php?f=41&t=15934&start=10#p171312
- NOTE: props needs mass value >0 for footprint and blockpath to work

- TROPICAL single ficus' and palm's remain untouched

4) READ CHANGED BP CSV & OVERRIDE CURRENT PROP BP (python)
- ensure right structure-tree for different prop types (e.g. rock, tree, structures & wreckage have different value requirements)
- fix NUMBER VALUES like ReclaimEnergyMax = "" --> 0
- ensure AUDIO for props that are trees or should behave like those e.g. streetlights, fences behave like UEF counterparts (except Cybran_Fence_prop.bp - too big)
- make remaining props 'RECLAIMABLE' and remove  'SELECTABLE' from Ice_Crystal_01_prop.bp  
- ensure COLLISION OFFSET for props that are trees or should behave like those e.g. streetlight, fences (except Cybran_Fence_prop.bp - too big)
- ensure DEFENSE for remaining props
- remove LOD4 from Brch01_prop.bp
- ensure BLOCK PATH (true, false) and FOOTPRINT (OccupancyCaps = 3, OccupancyCaps = 0) if needed. Both are connected and bound to ReclaimMassMax (>= 10 blocks, <10 does not block)
- DO NOT BLOCK PATH for floating or flat props even if mass value is > 10 (only used for ThetaBridge01_prop.bp, Cybran_Dock_prop.bp, UEF_Dock_prop.bp & Seraphim_Car_02_prop.bp)
- made geyser's, mudpot's and theta bridge MAP ELEMENTS (similar to FAF Seraphim II props): Set category to 'INVULNERABLE' (removed 'RECLAIMABLE'), reclaim values to 1 (only to ensure block path if needed) and LOD to 3000
- remove FootprintSizeX/Y => redundant, seems to be replaced by FOOTPRINT (only used in RedRocks\Props\Rock_SM01-7_prop.bp, RedRocks\Props\Rock01-2_prop.bp, Tropical\Props\Rocks\Rock01_prop.bp)
- ensure SCRIPTCLASS & SCRIPTMODULE for all props that are trees or should behave like those e.g. fence, streetlight

shalkya commented:
conflict with "advanced target priorities v1."
![image](https://user-images.githubusercontent.com/11618427/54087314-f4347000-4351-11e9-848b-4553901d416a.png)
if you put on "advanced target priorities v1." and "Additional Camera Stuff v3.2" you get a black screen.

badbwoi commented:
would be nice if you could do it, last rebase took me quite some time - gitnoob-
or tell my which branch and I will try to improve myself. seems to me that I rebased it on the wrong branch last time.

shalkya commented:
Actually i would prefer that you rebase, so we can test it on top of other changes.
The way to do it : 
on your repo :
git checkout deploy/fafdevelop
git pull --rebase (git pull should work too)
git checkout props-blueprints
git pull --rebase deploy/fafdevelop

The PR should be updated automatically

shalkya commented:
The transition in the look while zooming in the "hollow"/"Norfair" trees looks weird. Can it be smoother ?
burnt trees on "haven's reef" looks wrong too
![image](https://user-images.githubusercontent.com/11618427/54093232-ebfe2400-4395-11e9-9888-0eb57c92b51b.png)
in general why do we see the trees faster than the other reclaim ? I mean there is some high value reclaim that are seen later than the trees. (eg : "goodland")
On "funeral plains", trees zoomed view seems to disappear too fast.
looks good on a map like "farce in tha forest"
The light on "eye of the storm" doesn't worth anything and the truck on "haven's reef", gonna have to tag that onto balance
Will need the review of mappers too

badbwoi commented:
Thanks for testing und feedback. I also tested all the maps you named.

> 
> 
> The transition in the look while zooming in the "hollow"/"Norfair" trees looks weird. Can it be smoother ?
> burnt trees on "haven's reef" looks wrong too
> On "funeral plains", trees zoomed view seems to disappear too fast when zoomed in.

Problem child detected: The lava trees, desert trees & desert cactus are the most problematic tree props in FA. The LOD3 was missing for groups, maybe because of what we see now. Due to the Supreme props integration, visibility when zooming out should be much improved already, but I could try to play with shader to get a better result/transition. Since I have no idea what the different shader's are doing it's a lot of try and error to get a "good" result. I already spend some time to find the right shader and that's the best result so far. Would be nice if someone could explain the different shader's to me!?

>
>
> in general why do we see the trees faster than the other reclaim ? I mean there is some high value reclaim that are seen later than the trees. (eg : "goodland")

The new LOD values orient themselves on the original FA values, that's why I treated trees differently. FA has a LOD of 700, 750 & 1000 for trees and tree groups (new 750). While rocks had the steps 100, 200, 250, 400, 500 & 1000 (new 300, 400, 750). 
Goodlands: That's why the "Carotenoid" groups among the rocks disappear later.

>
>
> The light (Aeon_Fence_prop.bp) on "eye of the storm" doesn't worth anything and the truck on "haven's reef", gonna have to tag that onto balance

The reclaim value changes are considered as BUG FIXES: http://forums.faforever.com/viewtopic.php?f=41&t=15934&start=10#p171312
Eye of the Etorm: Aeon_Fence_prop.bp from 300 to 5 mass (now values are similar to UEF_Fence_prop.bp and UEF_Streetlight_01_prop.bp - UEF was role model for fences and streetlights)
haven reef fixed: UEF_Bus_prop.bp 300 to 10 mass, UEF_Car1_prop.bp 300 to 10 mass, UEF_truck_prop.bp 300 to 37,5 mass (value by size applied)

badbwoi commented:
Hey, I started work on fixing the problematic tree transition.
> The transition in the look while zooming in the "hollow"/"Norfair" trees looks weird. Can it be smoother ?
> burnt trees on "haven's reef" looks wrong too
> On "funeral plains", trees zoomed view seems to disappear too fast when zoomed in.
I researched all shader combinations for trees and found a better combination. But the problem is that the LOD2 fades out way to early. That's why I changed the shader and adapted the LOD in the first place.

Here is what I got so far:
 ![tree_transition](https://user-images.githubusercontent.com/28671151/54315456-70db7e80-45de-11e9-97a0-982d88cd60c5.JPG)
-I tried to close "visibility gap" between LOD2 and LOD3. Set LOD2 Cutoff to 60 (displays LOD3 after that point) to avoid LOD2 fade out
-Video is showing the props unchanged + added LOD3 on the _**left**_ and the new approach + added LOD3 on the _**right**_: https://tinyurl.com/y3fvozly (needs to be downloaded - sry)

What is your opinion?
I kinda like the progress for "Dead" Lava theme trees. But it would be nice to make the "Manzanita" from Desert theme just a little bit more smooth (but I believe its better already).
Do you know what could cause the fade it, since no other trees are showing this behaviour? Could a referenced .dds be the culprit?

I will try a few more things before I make the next push.

shalkya commented:
it's way better already ! so much more than when the trees blink off/on
will try to see if people can help you on the technical side

badbwoi commented:
Dear balance team / interested person,

I prepared the following file to start a discussion about the lovely reclaim value changes. The file displays the changes for each of the 57 props in question and the reason why it should be changed. Pictures are added to get a feel for size ratio.

https://tinyurl.com/yxdcvqxv
_Please enter your thoughts and return the file to me (svenni_badbwoi: forum, discord, slack).
Note:
-Changes are marked yellow
-300 mass props are marked pink 
-Only use filter, sorting will change the picture order._

These changes are "no cosmetic changes" like LOD! So far the mapping community seems to welcome the changes. The applied values match "mass_value A", except "D", special "M" and "map elements".
![fixed_reclaim_values](https://user-images.githubusercontent.com/28671151/54492103-ea67bb00-48c3-11e9-9c49-c854f6e70143.png) 
_I recommend launching the "fa_props" map to compare all props and themes: https://tinyurl.com/y9xwczzu_

I intend to make the reclaim value of a prop more intuitive and also enable map makers to use all the themes for there maps (e.g. make a crystal map without ending up with mad reclaim values). The general idea is to fix bugged props (different value as props of same size and type) and improve the 300 mass props (most -if not all- of these props seems to be unfinished, because they have a copy&paste bp from the largest rock prop in the game "/Evergreen/Props/Rocks/SeaRock02_prop.bp").
![SeaRock02_prop](https://user-images.githubusercontent.com/28671151/54492117-0cf9d400-48c4-11e9-8423-d7485f0cbcfd.JPG)
Maybe some prop themes should/could be treated differently, here are some ideas:
1) Props of the faction's and crystalline theme could seen as more valuable compared to rock props. Value could be worth twice as much as props of same size
2) Structures could become map element too. So no value needs to be applied and they simply block path. But its kinda strange, because we are able to reclaim buildings of the same size all day in game.

I'm open for suggestions and look forward to your feedback.

badbwoi commented:
Hey, here are the details on my improvements/changes.

**LOD changes**
The new LOD values orient themselves on the original FA values.
But instead of LOD 750 as maximum, I  took 1000 LOD for better visibility for normal props. In that process i raised the LOD for rocks = 10 mass fromm 400 to 500.
Current FA trees and tree groups have a LOD of 700, 750 & 1000 (new LOD 1000).
While current rocks have the steps 100, 200, 250, 400, 500 & 1000 (new LOD 300, 500 or 750).
-new 300 LOD = rocks/structures <10 Mass
-new 500 LOD = rocks/structures 10 Mass
-new 1000 LOD = rocks/structures >10 Mass


**reclaim changes & map elements**
I talked to jagged (balance team) and we came to an agreement concerning the reclaim. The changes are included.
-55 props were changed (including 19 map elements)
-map elements: LOD was set to 1000 & reclaim values has been set to 1 (energy, mass, time)
![crystalline_changes](https://user-images.githubusercontent.com/28671151/57193669-0a1f7500-6f3e-11e9-9e24-b77810baa489.png)
![geothermal_changes](https://user-images.githubusercontent.com/28671151/57193674-17d4fa80-6f3e-11e9-92fb-d770e5309d51.png)
![redrocks_changes](https://user-images.githubusercontent.com/28671151/57193678-215e6280-6f3e-11e9-9fae-790e5fbd8c02.png)
![seraphimii_changes](https://user-images.githubusercontent.com/28671151/57193680-24f1e980-6f3e-11e9-92b5-9b8413c7f97e.png)
![structures_changes](https://user-images.githubusercontent.com/28671151/57193682-2b806100-6f3e-11e9-8d9f-198266ff5a10.png)
![swamp_changes](https://user-images.githubusercontent.com/28671151/57193683-2e7b5180-6f3e-11e9-9e24-cf2af24384e7.png)
![factions_wreackge_changes](https://user-images.githubusercontent.com/28671151/57193685-30451500-6f3e-11e9-84b7-651a63e39c4a.png)


**collision box fixes**
I reworked all the collision boxes (depends on the prop size) to block the path correctly. In some cases the model origin has an offset, that's why I fixed them to.
-check collision box fixes via "ctrl+shift+alt+c"
-check blocked path via "ctrl+alt+o" (blocked path is forced to in game grid - "normal offset")
-see test map: https://tinyurl.com/y4atf67a


**origin fixes** 
In order to block the path correctly (fix collision box offset) and display the reclaim value in the props center, the model origin for corrected for 56 models. This excludes "ThetaBridge01" because the origin is way off and all maps containing it, need to be reworked if the origin is changed.
-check collision box fixes via "ctrl+shift+alt+c"
-check blocked path via "ctrl+alt+o" (blocked path is forced to in game grid - "normal offset")
-see test map: https://tinyurl.com/y4atf67a
![image](https://user-images.githubusercontent.com/28671151/57193621-49999180-6f3d-11e9-98f5-025457965b34.png)


**tree changes**
To improve visibility and transition between the new LOD3 and old LOD2 is rather difficult for the desert and lava trees. I suspect the .dds is missing the top level information, because the props can be seen from the side even if you cant see them from the top anymore.
In order to still improve visibility, I assigned the "VertexNormal" shader to LOD2 for desert and lava trees. I also replaced the lod1.scm model of the desert "Manzanita" trees with the slightly more detailed model lod0.scm. This is the best I could come up with, concerning the given conditions.

Maybe the .dds could be fixed, but I'm unable to do it.
![dds_missing_top_perspective](https://user-images.githubusercontent.com/28671151/57193475-ac8a2900-6f3b-11e9-8908-c9ca72034092.png)


**other fixes**
-added missing files to seraphim_streetlight under Common and fixed it so it can be used
![image](https://user-images.githubusercontent.com/28671151/57193423-21109800-6f3b-11e9-8908-ae370bd60ff9.png)

-fixed formerly broken blueprint of geothermal "GeoRock01". It can now be used for the first time.
![image](https://user-images.githubusercontent.com/28671151/57193406-ffafac00-6f3a-11e9-9d24-f2dc7ae56f7c.png)

badbwoi commented:
I happy to present the improved trees.
After ozonex gave it a try, he discovered that the (last few) MIP maps don't hold enough information in the alpha channel. Based on his work/information -thx- and a lot of try and error I managed to improve the following ones:
Desert: Manzanita01_s1_albedo.dds
Evergreen: eg_fern01_albedo.dds, Brch01_s1_albedo.dds, Pine07_albedo.dds
Lava: Dead01_s1_albedo.dds, LV_Dead01_s1_albedo.dds, XDead01_s1_albedo.dds
Swamp: sw_fern01_albedo.dds

I also changed the LOD 1 Cutoff for trees from 30 to 60 to achieve a better model transition. The LOD steps for trees are now the following:
60, 200, 1000

shalkya commented:
tested, all the request have been satisfied. Well done.

speed2CZ commented on env/Common/Props/Clutter01_prop.bp:
```diff
@@ -1,14 +1,35 @@
                 {
                     AlbedoName = 'Clutter01_albedo.dds',
-                    LODCutoff = 10,
+                    LODCutoff = 300,
```
I dont think there super small props needs such high LODCutoff
